File: backend/futureu/src/main/java/edu/cit/futureu/service/ProfileService.java
package edu.cit.futureu.service;

import java.io.IOException;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import edu.cit.futureu.entity.UserEntity;
import edu.cit.futureu.repository.UserRepository;

@Service
public class ProfileService {
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;
    
    @Autowired
    private InstitutionService institutionService;
    
    @Autowired
    private CloudinaryService cloudinaryService; // NEW: Inject Cloudinary service
    
    // REMOVED: private final String uploadDir = "uploads/profile-pictures/";
    
    public UserEntity getUserProfile(int userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));
    }
    
    public UserEntity updateUserProfile(int userId, UserEntity profileData) {
        try {
            System.out.println("=== UPDATE USER PROFILE ===");
            System.out.println("Updating user with ID: " + userId);
            
            Optional<UserEntity> userOpt = userRepository.findById(userId);
            if (!userOpt.isPresent()) {
                throw new RuntimeException("User not found with ID: " + userId);
            }
            
            UserEntity existingUser = userOpt.get();
            System.out.println("Found existing user: " + existingUser.getEmail());
            
            // Update fields
            if (profileData.getFirstName() != null) {
                existingUser.setFirstName(profileData.getFirstName());
            }
            if (profileData.getMiddleName() != null) {
                existingUser.setMiddleName(profileData.getMiddleName());
            }
            if (profileData.getLastname() != null) {
                existingUser.setLastname(profileData.getLastname());
            }
            if (profileData.getEmail() != null) {
                existingUser.setEmail(profileData.getEmail());
            }
            if (profileData.getAge() > 0) {
                existingUser.setAge(profileData.getAge());
            }
            if (profileData.getAddress() != null) {
                existingUser.setAddress(profileData.getAddress());
            }
            if (profileData.getContactNumber() != null) {
                existingUser.setContactNumber(profileData.getContactNumber());
            }
            
            // Handle school code for counselors
            if (profileData.getSchoolCode() != null) {
                String schoolCode = profileData.getSchoolCode().trim();
                if (!schoolCode.isEmpty()) {
                    // FIXED: Change validateSchoolCode to isValidSchoolCode
                    boolean isValid = institutionService.isValidSchoolCode(schoolCode);
                    if (!isValid) {
                        throw new RuntimeException("Invalid school code provided");
                    }
                    existingUser.setSchoolCode(schoolCode);
                } else {
                    existingUser.setSchoolCode(null);
                }
            }
            
            UserEntity savedUser = userRepository.save(existingUser);
            System.out.println("User profile updated successfully");
            return savedUser;
            
        } catch (RuntimeException e) {
            System.err.println("=== RUNTIME ERROR ===");
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.err.println("=== END RUNTIME ERROR ===");
            throw e;
        } catch (Exception e) {
            System.err.println("=== CRITICAL ERROR ===");
            System.err.println("Unexpected error: " + e.getMessage());
            e.printStackTrace();
            System.err.println("=== END CRITICAL ERROR ===");
            throw e;
        }
    }
    
    /**
     * Upload profile picture to Cloudinary
     */
    public String uploadProfilePicture(int userId, MultipartFile file) throws IOException {
        try {
            System.out.println("=== UPLOAD PROFILE PICTURE ===");
            System.out.println("Uploading for user ID: " + userId);
            
            // Get existing user
            Optional<UserEntity> userOpt = userRepository.findById(userId);
            if (!userOpt.isPresent()) {
                throw new RuntimeException("User not found with ID: " + userId);
            }
            
            UserEntity user = userOpt.get();
            String oldProfilePictureUrl = user.getProfilePictureUrl();
            
            // Delete old image from Cloudinary if exists
            if (oldProfilePictureUrl != null && !oldProfilePictureUrl.isEmpty()) {
                try {
                    cloudinaryService.deleteProfilePicture(oldProfilePictureUrl);
                    System.out.println("Deleted old profile picture from Cloudinary");
                } catch (IOException e) {
                    System.err.println("Failed to delete old image: " + e.getMessage());
                    // Continue with upload even if deletion fails
                }
            }
            
            // Upload new image to Cloudinary
            String imageUrl = cloudinaryService.uploadProfilePicture(file, userId);
            System.out.println("Uploaded new profile picture to Cloudinary: " + imageUrl);
            
            // Update user's profile picture URL in database
            user.setProfilePictureUrl(imageUrl);
            userRepository.save(user);
            
            System.out.println("Profile picture URL updated in database");
            return imageUrl;
            
        } catch (IOException e) {
            System.err.println("Error uploading profile picture: " + e.getMessage());
            throw e;
        }
    }
    
    /**
     * Delete profile picture from Cloudinary
     */
    public void deleteProfilePicture(int userId) {
        Optional<UserEntity> userOpt = userRepository.findById(userId);
        if (userOpt.isPresent()) {
            UserEntity user = userOpt.get();
            String profilePictureUrl = user.getProfilePictureUrl();
            
            // Delete from Cloudinary
            if (profilePictureUrl != null && !profilePictureUrl.isEmpty()) {
                try {
                    cloudinaryService.deleteProfilePicture(profilePictureUrl);
                    System.out.println("Deleted profile picture from Cloudinary");
                } catch (IOException e) {
                    System.err.println("Failed to delete image from Cloudinary: " + e.getMessage());
                    // Continue to remove URL from database even if Cloudinary delete fails
                }
            }
            
            // Remove URL from database
            user.setProfilePictureUrl(null);
            userRepository.save(user);
        }
    }

    public void changePassword(int userId, String currentPassword, String newPassword) {
        Optional<UserEntity> userOpt = userRepository.findById(userId);
        if (!userOpt.isPresent()) {
            throw new RuntimeException("User not found");
        }
        
        UserEntity user = userOpt.get();
        
        // Verify current password
        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new RuntimeException("Current password is incorrect");
        }
        
        // Update to new password
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
}

File: frontend/futureu-capstone/src/services/profileService.js
import apiClient from './api';
import authService from './authService';

class ProfileService {
  /**
   * Get user profile
   */
  async getUserProfile() {
    try {
      // Use authService to get current user ID
      const userId = authService.getCurrentUserId();
      
      if (!userId) {
        throw new Error('User not authenticated or user ID not found');
      }

      const response = await apiClient.get(`/profile/${userId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching user profile:', error);
      throw error;
    }
  }

  /**
   * Update user profile
   */
  async updateUserProfile(profileData) {
    try {
      // Use authService to get current user ID
      const userId = authService.getCurrentUserId();
      
      if (!userId) {
        throw new Error('User not authenticated or user ID not found');
      }

      const response = await apiClient.put(`/profile/${userId}`, profileData);
      return response.data;
    } catch (error) {
      console.error('Error updating user profile:', error);
      throw error;
    }
  }

  /**
   * Upload profile picture to Cloudinary
   * @param {File} file - The image file to upload
   * @returns {Promise<string>} - The Cloudinary URL of the uploaded image
   */
  async uploadProfilePicture(file) {
    try {
      console.log('ProfileService: Uploading profile picture...');
      
      // Use authService to get current user ID
      const userId = authService.getCurrentUserId();
      
      if (!userId) {
        throw new Error('User not authenticated or user ID not found');
      }

      // Create FormData for file upload
      const formData = new FormData();
      formData.append('file', file);

      // Upload to backend (which will upload to Cloudinary)
      const response = await apiClient.post(`/profile/${userId}/upload-picture`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
        // Add timeout for large file uploads
        timeout: 60000, // 60 seconds
      });

      console.log('ProfileService: Upload response:', response.data);

      // Backend should return the Cloudinary URL in profilePictureUrl field
      const imageUrl = response.data.profilePictureUrl || response.data.imageUrl || response.data;
      
      if (!imageUrl) {
        throw new Error('No image URL returned from server');
      }

      return imageUrl;
    } catch (error) {
      console.error('ProfileService: Upload error:', error);
      
      // Handle specific error cases
      if (error.response?.status === 413) {
        throw new Error('File size too large. Please choose a smaller image (max 10MB).');
      } else if (error.response?.status === 415) {
        throw new Error('Invalid file type. Please upload an image file (JPG, PNG, etc.).');
      } else if (error.code === 'ECONNABORTED') {
        throw new Error('Upload timeout. Please try again with a smaller file.');
      }
      
      throw error;
    }
  }

  /**
   * Delete profile picture from Cloudinary
   */
  async deleteProfilePicture() {
    try {
      console.log('ProfileService: Deleting profile picture...');
      
      // Use authService to get current user ID
      const userId = authService.getCurrentUserId();
      
      if (!userId) {
        throw new Error('User not authenticated or user ID not found');
      }

      const response = await apiClient.delete(`/profile/${userId}/profile-picture`);
      
      console.log('ProfileService: Delete response:', response.data);
      return response.data;
    } catch (error) {
      console.error('ProfileService: Delete error:', error);
      throw error;
    }
  }

  /**
   * Change user password
   */
  async changePassword(currentPassword, newPassword) {
    try {
      // Use authService to get current user ID
      const userId = authService.getCurrentUserId();
      
      if (!userId) {
        throw new Error('User not authenticated or user ID not found');
      }

      const response = await apiClient.put(`/profile/${userId}/change-password`, {
        currentPassword,
        newPassword
      });
      return response.data;
    } catch (error) {
      console.error('Error changing password:', error);
      throw error;
    }
  }
}

export default new ProfileService();